File: csharp/ApiRequests.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Net.Http;
using System.Threading.Tasks;
using Newtonsoft.Json;
using System.IO;

namespace StackWizard
{
    /// <summary>
    /// API Request Performance Test for C#.
    /// Uses HttpClient with async/await for efficient concurrent HTTP requests.
    /// Collects comprehensive metrics including response times, throughput, and percentiles.
    /// </summary>
    public class ApiRequests
    {
        private static readonly HttpClient httpClient = new HttpClient
        {
            Timeout = TimeSpan.FromSeconds(10)
        };

        /// <summary>
        /// Result of a single API request.
        /// </summary>
        public class RequestResult
        {
            public int Id { get; set; }
            public bool Success { get; set; }
            public double ResponseTimeMs { get; set; }
            public double Timestamp { get; set; }
            public string? Error { get; set; }
        }

        /// <summary>
        /// Comprehensive API performance metrics.
        /// </summary>
        public class ApiMetrics
        {
            [JsonProperty("language")]
            public string Language { get; set; } = "csharp";

            [JsonProperty("total_requests")]
            public int TotalRequests { get; set; }

            [JsonProperty("successful_requests")]
            public int SuccessfulRequests { get; set; }

            [JsonProperty("failed_requests")]
            public int FailedRequests { get; set; }

            [JsonProperty("total_time_seconds")]
            public double TotalTimeSeconds { get; set; }

            [JsonProperty("requests_per_second")]
            public double RequestsPerSecond { get; set; }

            [JsonProperty("response_times")]
            public ResponseTimes? ResponseTimesData { get; set; }

            [JsonProperty("timeseries")]
            public List<TimeseriesPoint> Timeseries { get; set; } = new List<TimeseriesPoint>();
        }

        public class ResponseTimes
        {
            [JsonProperty("min_ms")]
            public double MinMs { get; set; }

            [JsonProperty("max_ms")]
            public double MaxMs { get; set; }

            [JsonProperty("average_ms")]
            public double AverageMs { get; set; }

            [JsonProperty("median_ms")]
            public double MedianMs { get; set; }

            [JsonProperty("p95_ms")]
            public double P95Ms { get; set; }

            [JsonProperty("p99_ms")]
            public double P99Ms { get; set; }
        }

        public class TimeseriesPoint
        {
            [JsonProperty("timestamp")]
            public double Timestamp { get; set; }

            [JsonProperty("response_time_ms")]
            public double ResponseTimeMs { get; set; }
        }

        /// <summary>
        /// Make a single HTTP request asynchronously.
        /// </summary>
        private static async Task<RequestResult> MakeRequestAsync(string url, int requestId, Stopwatch totalWatch)
        {
            var requestWatch = Stopwatch.StartNew();
            try
            {
                var response = await httpClient.GetAsync(url);
                requestWatch.Stop();

                return new RequestResult
                {
                    Id = requestId,
                    Success = response.IsSuccessStatusCode,
                    ResponseTimeMs = requestWatch.Elapsed.TotalMilliseconds,
                    Timestamp = totalWatch.Elapsed.TotalSeconds
                };
            }
            catch (Exception ex)
            {
                requestWatch.Stop();
                return new RequestResult
                {
                    Id = requestId,
                    Success = false,
                    ResponseTimeMs = requestWatch.Elapsed.TotalMilliseconds,
                    Timestamp = totalWatch.Elapsed.TotalSeconds,
                    Error = ex.Message
                };
            }
        }

        /// <summary>
        /// Calculate percentile from sorted list.
        /// </summary>
        private static double Percentile(List<double> sortedList, double percentile)
        {
            int index = (int)Math.Ceiling(sortedList.Count * percentile) - 1;
            index = Math.Max(0, Math.Min(index, sortedList.Count - 1));
            return sortedList[index];
        }

        /// <summary>
        /// Run API request benchmark with 10,000 concurrent requests.
        /// </summary>
        public static async Task<ApiMetrics> RunBenchmarkAsync()
        {
            string url = "https://jsonplaceholder.typicode.com/posts/1";
            int numRequests = 1000;

            Console.WriteLine($"Starting benchmark: {numRequests} requests to {url}");

            var totalWatch = Stopwatch.StartNew();

            // Make concurrent requests using Task.WhenAll
            var tasks = Enumerable.Range(0, numRequests)
                                 .Select(i => MakeRequestAsync(url, i, totalWatch))
                                 .ToArray();

            var results = await Task.WhenAll(tasks);

            totalWatch.Stop();
            double totalTime = totalWatch.Elapsed.TotalSeconds;

            // Analyze results
            var successfulResults = results.Where(r => r.Success).ToList();
            var failedResults = results.Where(r => !r.Success).ToList();

            int successfulCount = successfulResults.Count;
            int failedCount = failedResults.Count;

            // Extract and sort response times
            var responseTimes = successfulResults.Select(r => r.ResponseTimeMs).ToList();
            var sortedTimes = responseTimes.OrderBy(t => t).ToList();

            // Create metrics
            var metrics = new ApiMetrics
            {
                TotalRequests = numRequests,
                SuccessfulRequests = successfulCount,
                FailedRequests = failedCount,
                TotalTimeSeconds = Math.Round(totalTime, 2),
                RequestsPerSecond = Math.Round(numRequests / totalTime, 2)
            };

            if (responseTimes.Count > 0)
            {
                metrics.ResponseTimesData = new ResponseTimes
                {
                    MinMs = Math.Round(responseTimes.Min(), 2),
                    MaxMs = Math.Round(responseTimes.Max(), 2),
                    AverageMs = Math.Round(responseTimes.Average(), 2),
                    MedianMs = Math.Round(Percentile(sortedTimes, 0.50), 2),
                    P95Ms = Math.Round(Percentile(sortedTimes, 0.95), 2),
                    P99Ms = Math.Round(Percentile(sortedTimes, 0.99), 2)
                };

                metrics.Timeseries = successfulResults.Select(r => new TimeseriesPoint
                {
                    Timestamp = Math.Round(r.Timestamp, 3),
                    ResponseTimeMs = Math.Round(r.ResponseTimeMs, 2)
                }).ToList();
            }

            // Save results to JSON file
            string outputFile = "api_results.json";
            string jsonData = JsonConvert.SerializeObject(metrics, Formatting.Indented);
            File.WriteAllText(outputFile, jsonData);

            Console.WriteLine($"\nResults saved to {outputFile}");

            return metrics;
        }

        /// <summary>
        /// Print formatted results to console.
        /// </summary>
        public static void PrintResults(ApiMetrics metrics)
        {
            Console.WriteLine("\n" + new string('=', 60));
            Console.WriteLine($"Language: {metrics.Language.ToUpper()}");
            Console.WriteLine(new string('=', 60));
            Console.WriteLine($"Total Requests: {metrics.TotalRequests}");
            Console.WriteLine($"Successful: {metrics.SuccessfulRequests}");
            Console.WriteLine($"Failed: {metrics.FailedRequests}");
            Console.WriteLine($"Total Time: {metrics.TotalTimeSeconds:F2}s");
            Console.WriteLine($"Requests/sec: {metrics.RequestsPerSecond:F2}");

            if (metrics.ResponseTimesData != null)
            {
                var rt = metrics.ResponseTimesData;
                Console.WriteLine("\nResponse Times (ms):");
                Console.WriteLine($"  Min: {rt.MinMs:F2}");
                Console.WriteLine($"  Max: {rt.MaxMs:F2}");
                Console.WriteLine($"  Avg: {rt.AverageMs:F2}");
                Console.WriteLine($"  Median: {rt.MedianMs:F2}");
                Console.WriteLine($"  P95: {rt.P95Ms:F2}");
                Console.WriteLine($"  P99: {rt.P99Ms:F2}");
            }

            Console.WriteLine(new string('=', 60) + "\n");
        }
    }
}

File: go/api_requests.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"
	"os"
	"path/filepath"
	"sort"
	"sync"
	"time"
)

// RequestResult holds the result of a single request
type RequestResult struct {
	ID             int     `json:"id"`
	Success        bool    `json:"success"`
	ResponseTimeMs float64 `json:"response_time_ms"`
	Timestamp      float64 `json:"timestamp"`
	Error          string  `json:"error,omitempty"`
}

// Metrics holds all benchmark metrics
type Metrics struct {
	Language           string                   `json:"language"`
	TotalRequests      int                      `json:"total_requests"`
	SuccessfulRequests int                      `json:"successful_requests"`
	FailedRequests     int                      `json:"failed_requests"`
	TotalTimeSeconds   float64                  `json:"total_time_seconds"`
	RequestsPerSecond  float64                  `json:"requests_per_second"`
	ResponseTimes      map[string]float64       `json:"response_times"`
	Timeseries         []map[string]float64     `json:"timeseries"`
}

// Make a single HTTP request with timing
func makeRequest(client *http.Client, url string, requestID int, startTime time.Time, results chan<- RequestResult, wg *sync.WaitGroup) {
	defer wg.Done()

	requestStart := time.Now()
	resp, err := client.Get(url)
	requestEnd := time.Now()

	result := RequestResult{
		ID:             requestID,
		ResponseTimeMs: float64(requestEnd.Sub(requestStart).Microseconds()) / 1000.0,
		Timestamp:      requestEnd.Sub(startTime).Seconds(),
	}

	if err != nil {
		result.Success = false
		result.Error = err.Error()
	} else {
		defer resp.Body.Close()
		io.ReadAll(resp.Body) // Read and discard body
		result.Success = resp.StatusCode == 200
	}

	results <- result
}

// Calculate percentile from sorted slice
func percentile(sortedData []float64, p float64) float64 {
	if len(sortedData) == 0 {
		return 0
	}
	index := int(math.Ceil(float64(len(sortedData)) * p)) - 1
	if index < 0 {
		index = 0
	}
	if index >= len(sortedData) {
		index = len(sortedData) - 1
	}
	return sortedData[index]
}

// RunAPIBenchmark runs API request benchmark with 1,000 concurrent requests
func RunAPIBenchmark() {
	url := "https://jsonplaceholder.typicode.com/posts/1"
	numRequests := 1000

	fmt.Printf("Starting benchmark: %d requests to %s\n", numRequests, url)

	// Create HTTP client with connection pooling
	client := &http.Client{
		Transport: &http.Transport{
			MaxIdleConnsPerHost: 50,
			MaxConnsPerHost:     50,
		},
		Timeout: 10 * time.Second,
	}

	// Measure total time for concurrent requests
	startTime := time.Now()

	var wg sync.WaitGroup
	results := make(chan RequestResult, numRequests)

	for i := 0; i < numRequests; i++ {
		wg.Add(1)
		go makeRequest(client, url, i, startTime, results, &wg)
	}

	wg.Wait()
	close(results)

	endTime := time.Now()
	totalTime := endTime.Sub(startTime).Seconds()

	// Analyze results
	var allResults []RequestResult
	var successfulResults []RequestResult
	var responseTimes []float64

	for result := range results {
		allResults = append(allResults, result)
		if result.Success {
			successfulResults = append(successfulResults, result)
			responseTimes = append(responseTimes, result.ResponseTimeMs)
		}
	}

	successfulCount := len(successfulResults)
	failedCount := len(allResults) - successfulCount

	// Calculate metrics
	metrics := Metrics{
		Language:           "go",
		TotalRequests:      numRequests,
		SuccessfulRequests: successfulCount,
		FailedRequests:     failedCount,
		TotalTimeSeconds:   math.Round(totalTime*100) / 100,
		RequestsPerSecond:  math.Round(float64(numRequests)/totalTime*100) / 100,
		ResponseTimes:      make(map[string]float64),
		Timeseries:         make([]map[string]float64, 0),
	}

	if len(responseTimes) > 0 {
		sort.Float64s(responseTimes)

		sum := 0.0
		minVal := responseTimes[0]
		maxVal := responseTimes[len(responseTimes)-1]
		for _, rt := range responseTimes {
			sum += rt
		}

		metrics.ResponseTimes = map[string]float64{
			"min_ms":     math.Round(minVal*100) / 100,
			"max_ms":     math.Round(maxVal*100) / 100,
			"average_ms": math.Round((sum/float64(len(responseTimes)))*100) / 100,
			"median_ms":  math.Round(percentile(responseTimes, 0.5)*100) / 100,
			"p95_ms":     math.Round(percentile(responseTimes, 0.95)*100) / 100,
			"p99_ms":     math.Round(percentile(responseTimes, 0.99)*100) / 100,
		}

		// Build timeseries data
		for _, result := range successfulResults {
			metrics.Timeseries = append(metrics.Timeseries, map[string]float64{
				"timestamp":        math.Round(result.Timestamp*1000) / 1000,
				"response_time_ms": math.Round(result.ResponseTimeMs*100) / 100,
			})
		}
	}

	// Save results to JSON file
	execPath, _ := os.Executable()
	execDir := filepath.Dir(execPath)
	outputFile := filepath.Join(execDir, "api_results.json")

	// If running with 'go run', use current directory
	if _, err := os.Stat(execDir + "/api_requests.go"); os.IsNotExist(err) {
		outputFile = "api_results.json"
	}

	jsonData, err := json.MarshalIndent(metrics, "", "  ")
	if err != nil {
		fmt.Printf("Error marshaling JSON: %v\n", err)
		return
	}

	err = os.WriteFile(outputFile, jsonData, 0644)
	if err != nil {
		fmt.Printf("Error writing file: %v\n", err)
		return
	}

	fmt.Printf("\nResults saved to %s\n", outputFile)

	// Print results
	printResults(metrics)
}

func printResults(metrics Metrics) {
	fmt.Println("\n" + string(make([]byte, 60)))
	for i := range make([]byte, 60) {
		fmt.Print("=")
	}
	fmt.Printf("\nLanguage: %s\n", "GO")
	for i := range make([]byte, 60) {
		fmt.Print("=")
	}
	fmt.Printf("\nTotal Requests: %d\n", metrics.TotalRequests)
	fmt.Printf("Successful: %d\n", metrics.SuccessfulRequests)
	fmt.Printf("Failed: %d\n", metrics.FailedRequests)
	fmt.Printf("Total Time: %.2fs\n", metrics.TotalTimeSeconds)
	fmt.Printf("Requests/sec: %.2f\n", metrics.RequestsPerSecond)

	if len(metrics.ResponseTimes) > 0 {
		fmt.Println("\nResponse Times (ms):")
		fmt.Printf("  Min: %.2f\n", metrics.ResponseTimes["min_ms"])
		fmt.Printf("  Max: %.2f\n", metrics.ResponseTimes["max_ms"])
		fmt.Printf("  Avg: %.2f\n", metrics.ResponseTimes["average_ms"])
		fmt.Printf("  Median: %.2f\n", metrics.ResponseTimes["median_ms"])
		fmt.Printf("  P95: %.2f\n", metrics.ResponseTimes["p95_ms"])
		fmt.Printf("  P99: %.2f\n", metrics.ResponseTimes["p99_ms"])
	}
	for i := range make([]byte, 60) {
		fmt.Print("=")
	}
	fmt.Println()
}

func main() {
	RunAPIBenchmark()
}
